controllers/openstackplaybookgenerator_controller.go

```diff
@@ -0,0 +1,352 @@
+	}
+	templateParameters["TripleoCustomDeployFiles"] = tripleoCustomDeployFiles
+
+	// get tripleo-deploy-config, created/rendered by openstackipset controller
+	tripleoDeployCM, _, err := common.GetConfigMapAndHashWithName(r, "tripleo-deploy-config", instance.Namespace)
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			return ctrl.Result{RequeueAfter: time.Second * 10}, err
+		}
+		return ctrl.Result{}, err
+	}
+
+	tripleoDeployFiles := tripleoDeployCM.Data
+	templateParameters["TripleoDeployFiles"] = tripleoDeployFiles
+	templateParameters["HeatServiceName"] = "heat-" + instance.Name
+	templateParameters["RolesFile"] = instance.Spec.RolesFile
+
+	// config hash
+	configMapHash, err := common.ObjectHash([]interface{}{tripleoCustomDeployCM.Data, tripleoDeployCM.Data})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("error calculating configmap hash: %v", err)
+	}
+
+	// extra stuff if custom tarballs are provided
+	if instance.Spec.HeatTemplateTarball != "" {
+		tripleoTarballCM, _, err := common.GetConfigMapAndHashWithName(r, instance.Spec.HeatTemplateTarball, instance.Namespace)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		// adjust the configMapHash
+		configMapHash, err = common.ObjectHash([]interface{}{tripleoCustomDeployCM.Data, tripleoDeployCM.Data, tripleoTarballCM.BinaryData})
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error calculating configmap hash: %v", err)
+		}
+		tripleoTarballFiles := tripleoTarballCM.BinaryData
+		templateParameters["TripleoTarballFiles"] = tripleoTarballFiles
+	}
+
+	cms = []common.Template{
+		{
+			Name:           "openstackplaybook-script-" + instance.Name,
+			Namespace:      instance.Namespace,
+			Type:           common.TemplateTypeScripts,
+			InstanceType:   instance.Kind,
+			AdditionalData: map[string]string{},
+			ConfigOptions:  templateParameters,
+			Labels:         cmLabels,
+		},
+	}
+	err = common.EnsureConfigMaps(r, instance, cms, &envVars)
+	if err != nil {
+		return ctrl.Result{}, nil
+	}
+
+	// Create ephemeral heat
+	heat := &ospdirectorv1beta1.OpenStackEphemeralHeat{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      instance.Name,
+			Namespace: instance.Namespace,
+		},
+		Spec: ospdirectorv1beta1.OpenStackEphemeralHeatSpec{
+			ConfigHash: configMapHash,
+		},
+	}
+
+	op, err := controllerutil.CreateOrUpdate(context.TODO(), r.Client, heat, func() error {
+		err := controllerutil.SetControllerReference(instance, heat, r.Scheme)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if op != controllerutil.OperationResultNone {
+		r.Log.Info(fmt.Sprintf("OpenStackEphemeralHeat successfully reconciled - operation: %s", string(op)))
+	}
+
+	// delete ephemeral heat if the configMapHash doesn't match
+	r.Log.Info(fmt.Sprintf("ConfigHash: %s", configMapHash))
+	r.Log.Info(fmt.Sprintf("HeatConfigHash: %s", heat.Spec.ConfigHash))
+	if heat.Spec.ConfigHash != configMapHash {
```

Comment:
This will trigger if `heat` resource already existed and the hashes are different, right?  I guess since we set the `ConfigHash` attribute on line 202 and not within the mutate function of the `CreateOrUpdate` call, then the `CreateOrUpdate` call will return a `heat` resource with a previous hash if a resource with `instance.Name` already existed.